src/qiskit_qir/visitor.py

```diff
@@ -86,64 +104,49 @@ def visit_instruction(self, instruction):
+    def __init__(self, profile: str = "AdaptiveExecution", **kwargs):
         self._module = None
         self._builder = None
+        self._qis = None
         self._qubit_labels = {}
         self._clbit_labels = {}
         self._profile = profile
         self._capabilities = self._map_profile_to_capabilities(profile)
         self._measured_qubits = {}
-        self._use_static_qubit_alloc = kwargs.get("use_static_qubit_alloc", True)
-        self._use_static_result_alloc = kwargs.get("use_static_result_alloc", True)
         self._emit_barrier_calls = kwargs.get("emit_barrier_calls", False)
         self._record_output = kwargs.get("record_output", True)
 
-    def visit_qiskit_module(self, module):
+    def visit_qiskit_module(self, module: QiskitModule):
         _log.debug(
             f"Visiting Qiskit module '{module.name}' ({module.num_qubits}, {module.num_clbits})"
         )
-        self._module = SimpleModule(
-            name=module.name,
-            num_qubits=module.num_qubits,
-            num_results=module.num_clbits,
+        self._module = module.module
+        context = self._module.context
+        entry = entry_point(
+            self._module, module.name, module.num_qubits, module.num_clbits
         )
-        self._module.use_static_qubit_alloc(self._use_static_qubit_alloc)
-        self._module.use_static_result_alloc(self._use_static_result_alloc)
+        # we update the parent module with the unique name
+        # created when defining in function in case of conflicts
+        module._entry_point = entry.name
+        self._builder = Builder(context)
+        self._builder.insert_at_end(BasicBlock(context, "entry", entry))
+        self._qis = BasicQisBuilder(self._builder)
 
-        self._builder = BasicQisBuilder(self._module.builder)
+    def finalize(self):
+        self._builder.ret(None)
 
-        self._barrier = self._module.add_external_function(
-            "__quantum__qis__barrier__body", types.Function([], types.VOID)
-        )
-        self._ccx = self._module.add_external_function(
-            "__quantum__qis__ccnot__body",
-            types.Function([types.QUBIT, types.QUBIT, types.QUBIT], types.VOID),
-        )
-        self._swap = self._module.add_external_function(
-            "__quantum__qis__swap__body",
-            types.Function([types.QUBIT, types.QUBIT], types.VOID),
-        )
-
-    def record_output(self, module):
+    def record_output(self, module: QiskitModule):
         if self._record_output == False:
             return
-
-        # produces output records of exactly "RESULT ARRAY_START"
-        array_start_record_output = self._module.add_external_function(
-            "__quantum__rt__array_start_record_output", types.Function([], types.VOID)
+        import pyqir.rt as rt
+        from pyqir import (
+            Constant,
+            IntType,
+            PointerType,
+            const,
```

Comment:
nit: Kinda weird to see local imports, is there a reason not to lift these to the top?